File: backend/src/test/java/lan/dk/podcastserver/manager/worker/gulli/GulliFinderTest.java
package lan.dk.podcastserver.manager.worker.gulli;

import com.github.davinkevin.podcastserver.service.HtmlService;
import com.github.davinkevin.podcastserver.service.ImageService;
import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.Podcast;
import com.github.davinkevin.podcastserver.IOUtils;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import java.io.IOException;
import java.net.URISyntaxException;

import static lan.dk.podcastserver.assertion.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.anyString;
import static org.mockito.Mockito.when;

/**
 * Created by kevin on 13/10/2016
 */
@RunWith(MockitoJUnitRunner.class)
public class GulliFinderTest {

    @Mock HtmlService htmlService;
    @Mock ImageService imageService;
    @InjectMocks
    GulliFinder gulliFinder;

    @Test
    public void should_find_a_podcast_for_valid_url() throws IOException, URISyntaxException {
        /* Given */
        when(htmlService.get(anyString())).thenReturn(IOUtils.fileAsHtml("/remote/podcast/gulli/pokemon.html"));
        when(imageService.getCoverFromURL(anyString())).then(i -> Cover.builder().url(i.getArgument(0)).height(250).width(250).build());

        /* When */
        Podcast podcast = gulliFinder.find("http://replay.gulli.fr/dessins-animes/Pokemon3");

        /* Then */
        assertThat(podcast)
                .hasUrl("http://replay.gulli.fr/dessins-animes/Pokemon3")
                .hasTitle("Pokémon")
                .hasType("Gulli");

        assertThat(podcast.getCover())
                .hasUrl("http://resize1-gulli.ladmedia.fr/r/340,255,smartcrop,center-top/img/var/storage/imports/replay/images_programme/pokemon_s19.jpg")
                .hasHeight(250)
                .hasWidth(250);
    }

    @Test
    public void should_be_compatible() {
        assertThat(gulliFinder.compatibility("http://replay.gulli.fr/dessins-animes/Pokemon3"))
                .isEqualTo(1);
    }

    @Test
    public void should_not_be_compatible() {
        assertThat(gulliFinder.compatibility("http://foo.bar.fr/dessins-animes/Pokemon3"))
                .isGreaterThan(1);
    }
}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/manager/worker/gulli/GulliFinderTest.kt
package com.github.davinkevin.podcastserver.manager.worker.gulli

import com.github.davinkevin.podcastserver.IOUtils.fileAsHtml
import com.github.davinkevin.podcastserver.entity.Cover
import com.github.davinkevin.podcastserver.service.HtmlService
import com.github.davinkevin.podcastserver.service.ImageService
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension

/**
 * Created by kevin on 13/10/2016
 */
@ExtendWith(MockitoExtension::class)
class GulliFinderTest {

    @Mock lateinit var htmlService: HtmlService
    @Mock lateinit var imageService: ImageService
    @InjectMocks lateinit var finder: GulliFinder

    @Test
    fun `should_find a podcast for valid url`() {
        /* Given */
        val podcastUrl = "http://replay.gulli.fr/dessins-animes/Pokemon3"
        val coverUrl = "http://resize1-gulli.ladmedia.fr/r/340,255,smartcrop,center-top/img/var/storage/imports/replay/images_programme/pokemon_s19.jpg"

        whenever(htmlService.get(podcastUrl)).thenReturn(fileAsHtml("/remote/podcast/gulli/pokemon.html"))
        whenever(imageService.getCoverFromURL(coverUrl))
                .then { Cover().apply { url = it.getArgument(0); height = 250; width = 250 } }

        /* When */
        val podcast = finder.find(podcastUrl)

        /* Then */
        assertThat(podcast.url).isEqualTo(podcastUrl)
        assertThat(podcast.title).isEqualTo("Pokémon")
        assertThat(podcast.type).isEqualTo("Gulli")

        val cover = podcast.cover!!
        assertThat(cover.url).isEqualTo(coverUrl)
        assertThat(cover.height).isEqualTo(250)
        assertThat(cover.width).isEqualTo(250)
    }

    @Test
    fun `should be compatible`() {
        assertThat(finder.compatibility("http://replay.gulli.fr/dessins-animes/Pokemon3"))
                .isEqualTo(1)
    }

    @Test
    fun `should not be compatible`() {
        assertThat(finder.compatibility("http://foo.bar.fr/dessins-animes/Pokemon3"))
                .isEqualTo(Integer.MAX_VALUE)
    }
}
